src/build/app/view.rs

```diff
@@ -0,0 +1,103 @@
+//! A shim layer that allows querying certain info about `App`,
+//! despite the corresponding fields being private.
+
+use super::*;
+use crate::INTERNAL_ERROR_MSG;
+
+#[allow(missing_debug_implementations)]
+pub struct AppView<'app, 'help> {
```

Comment:
I don't see the motivation of doing this. Having extra struct just to avoid making it private, feels weird to me

Reply:
Maybe we could add getter methods to `App` and `Arg` which should solve this situation.

Reply:
I considered a few other alternative ways to implement interpolation and ended up with a separate struct as the one that solves all the problems at once.

The problems:

* Exposing all fields as `pub` is not future compatible because of pattern matching. 

  Imagine the following scenario: some user downstream commits something like 
  ```rust
  let clap::Arg { field1, field2 } = arg
  ```
  in their codebase, and in the next patch version we add `field3`. Boom.
  
  I've actually seen this in the wild; here's one recent example https://github.com/rust-cli/human-panic/issues/71. 
  
  For the Great Justice: this can be resolved by adding a private dummy field so users would be required to use the wildcard.

* Some of the fields have private types, like ArgFlags (which is a bitfield). Exposing those fields would make expose the private types as well.

* Adding getters directly to `App/Arg` is totally fine, but... just look at the [documentation](https://docs.rs/clap/2.33.0/clap/struct.App.html): is already lengthily enough. I hope to move the methods to a separate page, that's it. If you both disagree, we can go with `get_` methods instead.

Reply:
I think we should try the `get_` methods first. Maybe we could use a trait to document them separately.

Reply:
> in their codebase, and in the next patch version we add field3. Boom.

Isn't this is what `#[non_exhaustive]` solve?

Reply:
> Isn't this is what `#[non_exhaustive]` solve?

Right, forgot it was stabilized. 

> like those in std,

Clarify?

Reply:
As in those patterns to get items, we do `x.item()` and to get a mutable item, we do `x.item_mut() = item`. Comparing it to the other languages, I rarely see the pattern `x.get_item()` and `x.set_item(item)` (except for the builder pattern which is used very often in rust) both together. Just my personal opinion on this based on what I saw from the existing API.

An example and clearer description could be seen in rust API guidelines https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

Reply:
The thing is that `x.item()` already serve as setters, so getters must have a prefix, exactly because `Arg & App` API is the builder pattern.

> we could hide it like how serde and other repositories have an internal API that's not exposed.

And this is the current state of affairs. It results in having a good half of the crate `pub hidden` (I'm not even exaggerating here, just grep for `#[doc(hidden)]`) . Getters are the attempt to avoid it and explicitly separate public and private API, our motivation.

Reply:
What if we could reuse the same builder API but for get too? I wonder if that's possible.